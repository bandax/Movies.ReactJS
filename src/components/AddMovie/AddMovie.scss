.add-movie {
    padding: 20px 0;
}

.btn-add-movie {
    background-color: rgba(10, 10, 10, 0.9);    
    font-size: 16px;
    padding: 10px 30px;    
    border-radius: 5px;    
    text-transform: uppercase; 
    border-color: #232323;    
}
