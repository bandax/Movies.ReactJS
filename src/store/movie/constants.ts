export const LOADING_MOVIES = "LOADING_MOVIES";
export type LOADING_MOVIES = typeof LOADING_MOVIES;

export const LOAD_MOVIES_SUCCESS = "LOAD_MOVIES_SUCCESS";
export type LOAD_MOVIES_SUCCESS = typeof LOAD_MOVIES_SUCCESS;

export const LOADING_MOVIES_ERROR = "LOADING_MOVIES_ERROR";
export type LOADING_MOVIES_ERROR = typeof LOADING_MOVIES_ERROR;

export const SELECT_MOVIE = "SELECT_MOVIE";
export type SELECT_MOVIE = typeof SELECT_MOVIE;

export const ADD_MOVIE = "ADD_MOVIE";
export type ADD_MOVIE = typeof ADD_MOVIE;

export const UPDATE_MOVIE = "UPDATE_MOVIE";
export type UPDATE_MOVIE = typeof UPDATE_MOVIE;

export const DELETE_MOVIE = "DELETE_MOVIE";
export type DELETE_MOVIE = typeof DELETE_MOVIE;
