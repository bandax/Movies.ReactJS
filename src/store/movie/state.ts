import { IMovie } from "../../interfaces/IResultsMovies";

export interface MovieState {
  movies: IMovie[];
  loading: boolean;
  errorMessage: string;
}
