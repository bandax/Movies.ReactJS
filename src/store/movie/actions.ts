import * as constants from "./constants";
import { IMovie } from "../../interfaces/IResultsMovies";
import { MovieActionsTypes } from "./types";

export function loadingMovies(): MovieActionsTypes {
  return {
    type: constants.LOADING_MOVIES,
    loading: true,
  };
}

export function loadingMoviesSuccess(movies: IMovie[]): MovieActionsTypes {
  return {
    type: constants.LOAD_MOVIES_SUCCESS,
    loading: false,
    movies: movies,
  };
}

export function loadingMoviesError(errorMessage: string): MovieActionsTypes {
  return {
    type: constants.LOADING_MOVIES_ERROR,
    loading: true,
    errorMessage: errorMessage,
  };
}
export function addMovie(movie: IMovie): MovieActionsTypes {
  return {
    type: constants.ADD_MOVIE,
    movieToAdd: movie,
  };
}

export function updateMovie(movie: IMovie): MovieActionsTypes {
  return {
    type: constants.UPDATE_MOVIE,
    movieToUpdate: movie,
  };
}

export function deleteMovie(movieId: number): MovieActionsTypes {
  return {
    type: constants.DELETE_MOVIE,
    movieId: movieId,
  };
}
