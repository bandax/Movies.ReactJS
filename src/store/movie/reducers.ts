import { MovieActionsTypes } from './types';
import { MovieState } from './state';
import { ACTIONS } from './constants';
import { IMovieData } from '../../interfaces/IMovieData';

const initialState: MovieState = {
  movies: [],
  loading: false,
  errorMessage: '',
  findMovies: true,
  genre: 'All',
};

export function movieReducer(
  state: MovieState = initialState,
  action: MovieActionsTypes,
): MovieState {
  switch (action.type) {
    case ACTIONS.ADD_MOVIE: {
      const addMovie = [...state.movies, action.movieToAdd];
      return {
        ...state,
        movies: addMovie,
      };
    }
    case ACTIONS.UPDATE_MOVIE: {
      const index = state.movies.findIndex(
        (movie) => movie.id === action.movieToUpdate.id,
      );
      const updatedMovies = [...state.movies];
      updatedMovies[index] = action.movieToUpdate;
      return {
        ...state,
        movies: updatedMovies,
      };
    }
    case ACTIONS.DELETE_MOVIE: {
      return {
        ...state,
        movies: state.movies.filter((movie) => movie.id !== action.movieId),
      };
    }
    case ACTIONS.LOADING_MOVIES: {
      return {
        ...state,
        loading: true,
      };
    }
    case ACTIONS.LOAD_MOVIES_SUCCESS: {
      let moviesToReturn = action.movies;
      if (state.genre.toLowerCase() !== 'all') {
        moviesToReturn = action.movies.filter((m) =>
          m.genres
            .map((g) => g.toLowerCase())
            .includes(state.genre.toLowerCase()),
        );
      }
      return {
        ...state,
        loading: false,
        movies: moviesToReturn,
        findMovies: action.movies.length > 0,
      };
    }
    case ACTIONS.LOADING_MOVIES_ERROR: {
      return {
        ...state,
        loading: false,
        errorMessage: action.errorMessage,
      };
    }
    case ACTIONS.SELECT_MOVIE: {
      return {
        ...state,
        loading: false,
        movie: action.movie,
      };
    }
    case ACTIONS.FILTER_BY_RELEASE_DATE_AND_RATING: {
      return {
        ...state,
        movies: state.movies.filter(
          (movie) =>
            movie.release_date === action.releaseDate &&
            action.rating === action.rating,
        ),
      };
    }
    case ACTIONS.SORT_BY_GENRE: {
      return {
        ...state,
        movies: state.movies.sort(compare),
      };
    }
    default:
      return state;
  }
}

function compare(a: IMovieData, b: IMovieData) {
  // Use toUpperCase() to ignore character casing
  const genresA = a.genres.map((genre) => genre.toUpperCase()).join(', ');
  const genresB = b.genres.map((genre) => genre.toUpperCase()).join(', ');

  if (genresA > genresB) {
    return 1;
  }
  if (genresA < genresB) {
    return -1;
  }
  return 0;
}
