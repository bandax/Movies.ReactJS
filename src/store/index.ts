import { combineReducers, applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { movieReducer } from './movie/reducers';

const rootReducer = combineReducers({
  movie: movieReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export const store = createStore(rootReducer, applyMiddleware(thunk));
