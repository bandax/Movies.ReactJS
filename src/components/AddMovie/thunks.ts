import { Action } from "redux";
import {
  addMovie,
  updateMovie,
  selectMovie,
  deleteMovie,
  loadingMoviesError,
} from "../../store/movie/actions";
import { RootState } from "../../store/index";
import { ThunkDispatch } from "redux-thunk";
import { IMovieData } from "../../interfaces/IMovieData";

export const saveMovie = (movie: IMovieData) => async (
  dispatch: ThunkDispatch<RootState, void, Action>
) => {
  try {
    let method = "post";
    if (movie.id === 0) {
      delete movie.id;
    } else {
      method = "put";
    }
    const body = JSON.stringify(movie);
    const response = await fetch("http://localhost:4000/movies", {
      headers: {
        "Content-Type": "application/json",
      },
      method: method,
      body,
    });

    if (response.status >= 200) {
      const movieUpdated = await response.json();
      if (method === "post") {
        dispatch(addMovie(movieUpdated));
      } else {
        dispatch(updateMovie(movieUpdated));
      }
    } else {
      const data = await response.json();
      const messages = data.messages;
      console.log(messages);

      let allMessages = "Error with message: \n";
      messages.forEach((message: string) => {
        allMessages = allMessages.concat(message + "\n");
      });
      alert(allMessages);
    }
  } catch (e) {
    dispatch(loadingMoviesError(e));
  }
};

export const deleteSelectedMovie = (id: number) => async (
  dispatch: ThunkDispatch<RootState, void, Action>
) => {
  try {
    const response = await fetch(`http://localhost:4000/movies/${id}`, {
      headers: {
        "Content-Type": "application/json",
      },
      method: "delete",
    });
    if (response.status >= 200) {
      dispatch(deleteMovie(id));
    }
  } catch (e) {
    dispatch(loadingMoviesError(e));
  }
};

export const selectedMovie = (movie: IMovieData) => (
  dispatch: ThunkDispatch<RootState, void, Action>
) => {
  dispatch(selectMovie(movie));
};
