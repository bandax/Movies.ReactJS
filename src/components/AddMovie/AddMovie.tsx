import * as React from 'react';
//import './AddMovie.scss';
import { AddMovieWindow } from '../AddMovieWindow/AddMovieWindow';
import clasificationTypes from '../../data/clasifications.json';
import { IMovieData } from '../../interfaces/IMovieData';
import { IClasification } from '../../interfaces/IClasificationMovie';
import { Action } from 'redux';

import { RootState } from '../../store/index';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import { saveMovie, selectedMovie } from './thunks';

const mapStateToProps = (state: RootState) => ({
  movie: state.movie.movie,
});

const mapDispatchToProps = (
  dispatch: ThunkDispatch<RootState, void, Action>
) => {
  return {
    addNewMovie: (movie: IMovieData) => dispatch(saveMovie(movie)),
    selectMovie: () => dispatch(selectedMovie(null)),
  };
};

type Props = ReturnType<typeof mapStateToProps> &
  ReturnType<typeof mapDispatchToProps>;

const AddMovie: React.FunctionComponent<Props> = ({
  movie,
  addNewMovie,
  selectMovie,
}) => {
  const [showAddMovieModal, setShowAddMovieModal] = React.useState<boolean>(
    false
  );
  const [movieTypes, setClasificationTypes] = React.useState<IClasification[]>(
    clasificationTypes
  );

  const handleShowAddMovieWindow = React.useCallback(() => {
    selectMovie();
    setShowAddMovieModal(!showAddMovieModal);
  }, [showAddMovieModal]);

  const handleSubmitAddMovie = function (movieToSave: IMovieData) {
    addNewMovie(movieToSave);
  };

  return (
    <>
      <div className="add-movie col-2">
        <button
          className="btn btn-add-movie"
          onClick={handleShowAddMovieWindow}
        >
          +Add Movie
        </button>
      </div>
      {/* <AddMovieWindow
        clasificationMovies={movieTypes}
        showModal={showAddMovieModal}
        movie={movie}
        onShowAddMovieWindow={handleShowAddMovieWindow}
        onAddMovieSubmit={handleSubmitAddMovie}
      /> */}
    </>
  );
};

export default connect(mapStateToProps, mapDispatchToProps)(AddMovie);
