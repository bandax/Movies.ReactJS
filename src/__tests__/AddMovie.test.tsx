import * as React from 'react';
import AddMovie from '../components/AddMovie/AddMovie';
import thunk, { ThunkDispatch } from 'redux-thunk';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import configureMockStore, { MockStoreCreator } from 'redux-mock-store';
import { MovieState } from '../store/movie/state';
import { Action } from 'redux';
import { Provider } from 'react-redux';
import { RootState } from '../store/index';

type DispatchExts = ThunkDispatch<MovieState, void, Action>;
const middleware = [thunk];
const state: RootState = {
  movie: {
    movie: null,
    errorMessage: '',
    findMovies: true,
    loading: false,
    movies: [],
  },
};

const mockStore = configureMockStore<RootState, DispatchExts>(middleware);

describe('<AddMovie />', () => {
  let store;

  beforeEach(() => (store = mockStore(state)));

  it('render a Add Movie Component', async () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AddMovie />
      </Provider>,
    );

    const fragment = asFragment();

    expect(fragment).toMatchSnapshot();
  });

  it('click to show Add Movie Window', async () => {
    const { getByText } = render(
      <Provider store={store}>
        <AddMovie />
      </Provider>,
    );
    const button = getByText('+Add Movie');

    userEvent.click(button);

    expect(getByText('Add movie window')).toBeInTheDocument();
  });
});
