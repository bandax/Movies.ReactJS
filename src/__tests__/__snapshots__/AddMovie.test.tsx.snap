// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddMovie /> render a Add Movie Component 1`] = `
<DocumentFragment>
  <div
    class="add-movie col-2"
  >
    <button
      class="btn btn-add-movie"
    >
      +Add Movie
    </button>
  </div>
</DocumentFragment>
`;
